lib/AST/ASTPrinter.cpp

```diff
@@ -506,35 +541,65 @@ class PrintAST : public ASTVisitor<PrintAST> {
                   SubstFlags::DesugarMemberTypes | SubstFlags::UseErrorType);
     }
 
-    printType(T);
+    printType(T, printCallback);
   }
 
-  void printTypeLocWithOptions(const TypeLoc &TL, PrintOptions options) {
+  void printTypeLoc(const TypeLoc &TL, TypePrintCallback printCallback = printTypeUnadorned) {
     if (CurrentType && TL.getType()) {
-      printTransformedType(TL.getType());
+      printTransformedType(TL.getType(), printCallback);
       return;
     }
 
     // Print a TypeRepr if instructed to do so by options, or if the type
     // is null.
-    if (willUseTypeReprPrinting(TL, CurrentType, options)) {
+    if (willUseTypeReprPrinting(TL, CurrentType, Options)) {
       if (auto repr = TL.getTypeRepr()) {
-        llvm::SaveAndRestore<bool> SPTA(options.SkipParameterTypeAttributes,
+        llvm::SaveAndRestore<bool> SPTA(Options.SkipParameterTypeAttributes,
                                         true);
-        repr->print(Printer, options);
+        repr->print(Printer, Options);
       }
       return;
     }
 
-    TL.getType().print(Printer, options);
+    printCallback(TL.getType(), Printer, Options);
   }
 
-  void printTypeLoc(const TypeLoc &TL) { printTypeLocWithOptions(TL, Options); }
+  static void printWithParensIfNotSimple(Type ty, ASTPrinter &printer, PrintOptions &opts) {
+    if (!isSimple(ty)) {
+      printer << "(";
+      ty.print(printer, opts);
+      printer << ")";
+    } else {
+      ty.print(printer, opts);
+    }
+  }
 
   void printTypeLocForImplicitlyUnwrappedOptional(TypeLoc TL) {
-    PrintOptions options = Options;
-    options.PrintOptionalAsImplicitlyUnwrapped = true;
-    printTypeLocWithOptions(TL, options);
+    // Unwrap the underlying type and print that, followed by a
+    // '!'.
+
+    if (auto *refTy = TL.getType()->getAs<ReferenceStorageType>()) {
+      // FIXME: This is a bit of a hack, but seems like the lesser of
+      // various evils. Ideally we would just create a new
+      // ReferenceStorageType with the underlying type of the
+      // optional, but we do not allow that for WeakStorageType, and
+      // removing that assertion would clearly be worse.
```

Comment:
Since we only print the attribute part here, why not separate that out into its own function instead? printReferenceTypeIntroducer and then go on to do the rest of the IUO stuff.